torch/_inductor/codecache.py

```diff
@@ -761,25 +761,51 @@ def get_include_and_linking_paths(
         ipaths = cpp_extension.include_paths(cuda) + [sysconfig.get_path("include")]
         lpaths = []
         if sys.platform == "darwin":
-            # GNU OpenMP generally is not available on MacOS
-            # There is either Intel OpenMP(for x86) or LLVM OpenMP (for both x86 and arm64)
-            libs = ["omp"]
-            if os.getenv("OMP_PREFIX") is not None:
-                # Support OpenMP on MacOS
+            cxx = cpp_compiler()
+            version_string = subprocess.check_output([cxx, "--version"]).decode('utf8')
+            # only Apple builtin compilers (Apple Clang++) require openmp
+            omp_available = 'Apple' not in version_string.splitlines()[0]
+
+            libs = [] if omp_available else ["omp"]
+
+            # prefer to use openmp from `conda install llvm-openmp`
+            if not omp_available and os.getenv("CONDA_PREFIX") is not None:
+                command = "conda list llvm-openmp --json"
+                output = subprocess.check_output(command.split()).decode('utf8')
+                omp_available = len(json.loads(output)) > 0
+                if omp_available:
+                    conda_lib_path = os.path.join(os.getenv("CONDA_PREFIX"), "lib")
+                    ipaths.append(os.path.join(os.getenv("CONDA_PREFIX"), "include"))
+                    lpaths.append(conda_lib_path)
+                    # Prefer Intel OpenMP on x86 machine
+                    if os.uname().machine == "x86_64" and os.path.exists(
+                        os.path.join(conda_lib_path, "libiomp5.dylib")
+                    ):
+                        libs = ["iomp5"]
+
+            # next, try to use openmp from `brew install libomp`
+            if not omp_available:
+                try:
+                    # check if `brew` is installed
+                    subprocess.check_output(["which", "brew"])
+                    # get the location of `libomp` if it is installed
+                    # this is the location that `libomp` **would** be installed
+                    # see https://github.com/Homebrew/brew/issues/10261#issuecomment-756563567 for details
+                    libomp_path = subprocess.check_output(["brew", "--prefix", "libomp"]).decode("utf8").strip()
+                    # check if `libomp` is installed
+                    omp_available = os.path.exists(libomp_path)
+                    if omp_available:
+                        ipaths.append(os.path.join(libomp_path, "include"))
+                        lpaths.append(os.path.join(libomp_path, "lib"))
+                except subprocess.SubprocessError:
+                    pass
+
+            if not omp_available and os.getenv("OMP_PREFIX") is not None:
                 ipaths.append(os.path.join(os.getenv("OMP_PREFIX"), "include"))
                 lpaths.append(os.path.join(os.getenv("OMP_PREFIX"), "lib"))
```

Comment:
Shall we unconditionally check the `OMP_PREFIX` as the original code does? This follows the user's specific intention.

Reply:
I think `OMP_PREFIX` is not user-friendly. Even after I `brew install libomp`, I still don't know how to set the `OMP_PREFIX` properly. It would be better if we can automatically discover the path, either via `CONDA_PREFIX` or via `brew --prefix libomp`. They would greatly help users.

Reply:
That's a good point. My point is that if we allow this option, we'd better check it first. Otherwise, we just remove this user-unfriendly option considering you have already improved the way of finding and configuring the omp header/library in this PR.

Reply:
We'd better keep this option to avoid breaking existing code. This also allows users who use neither `conda` nor `brew` to specify the location.

Sure, we can check it first. Fixed in [dff268b](https://github.com/pytorch/pytorch/pull/107111/commits/dff268bad9c1fdba8fa14c8f005b61b9bf4945ec)